### ISSUE TITLE
Be strict about strings in templates

### ISSUE THREAD

**Issue by silbinarywolf** [silverstripe/silverstripe-template-engine#20]
<details>
<summary>Original issue description</summary>

To kickstart the conversation properly, I'd like the ability to start making templates stricter.

Enabling this would do potentially do two things:
- Stop using identifiers as strings. (ie. you must quote your strings in *.ss templates)
- If a variable isn't explicitly set, an error will be thrown. This is a big deal to me as I spent 2h~ the other day tracking down a bug caused by the refactoring/moving of a function.

To begin, I think being able to decide what modules/themes can and can't be strict would be nice, Like below:
```php
$enable_strict = ['themes/mysite/MyTemplate.ss']
```
```php
$enable_strict = ['themes/mysite/*']
```

Example PoC:
https://github.com/kinglozzer/silverstripe-stricttemplates

</details>

Unquoted values are usually interpreted as strings - unless they're evaluated as a boolean, null, integer or float, or unless they are preceded with a `$`. This is referred to in the parser as a `FreeString` (as opposed to `QuotedString`).

This precludes us from parsing out constants, arrays, and any other types we may want that aren't just raw strings.

We should stop parsing unquoted strings as strings in arguments and control blocks. If it's quoted, it should be a string - otherwise it should be a constant (if it isn't interpreted as one of the other types).

Note that this is a breaking change, so would either need a feature flag or be implemented in a major release.

**Comment by dhensby** [comment]
> If a variable isn't explicitly set, an error will be thrown. This is a big deal to me as I spent 2h~ the other day tracking down a bug caused by the refactoring/moving of a function.

I don't think this is as important as the strict syntax. SS's templates don't have variables explicitly passed to them, so how would we determine an "explicitly set" variable? Template variables can come from a method, property or passed in explicitly. How do we determine if a var is "explicitly set" in a control loop? This just seems too tricky.

> being able to decide what modules/themes can and can't be strict would be nice

This would be _nice_, but isn't something I think should be a priority. If we add a flag we make our parser more complex and provide more opportunities for bugs, we also increase our maintenance burden.

My suggestion would be to make the parser stricter in terms of typing and inference of variables. unquoted strings become constants, quoted strings are strings and `$` prefixed strings are variables.

**Comment by tractorcow** [comment]
>  My suggestion would be to make the parser stricter in terms of typing and inference of variables. unquoted strings become constants, quoted strings are strings and $ prefixed strings are variables.

This in itself would add a lot of value. :)
